subsys/net/lib/zstream/zstream_tls.c

```diff
@@ -0,0 +1,228 @@
+
+static ssize_t zstream_tls_read(zstream stream, void *buf, size_t size)
+{
+	struct zstream_tls *self = (struct zstream_tls *)stream;
+	int ret = mbedtls_ssl_read(&self->ssl, buf, size);
+
+	if (ret >= 0) {
+		return ret;
+	}
+
+	if (ret == MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY) {
+		/* End of stream */
+		return 0;
+	}
+
+	if (ret == MBEDTLS_ERR_SSL_CLIENT_RECONNECT) {
+		/* We don't support TLS reconnects over the same socket,
+		 * treat as EOF.
+		 */
+		return 0;
+	}
+
+	if (ret == MBEDTLS_ERR_SSL_WANT_READ
+	    || ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
+		errno = EAGAIN;
+		return -1;
+	}
+
+	/* TODO: More error code conversion? */
+printf("mbedtls_ssl_read error: -%x\n", -ret);
+	errno = EINVAL;
+	return -1;
+}
+
+static ssize_t zstream_tls_write(zstream stream, const void *buf, size_t size)
+{
+	struct zstream_tls *self = (struct zstream_tls *)stream;
+	int ret = mbedtls_ssl_write(&self->ssl, buf, size);
+
+	if (ret >= 0) {
+		return ret;
+	}
+
+	if (ret == MBEDTLS_ERR_SSL_WANT_READ
+	    || ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
+		errno = EAGAIN;
+		return -1;
+	}
+
+	/* TODO: More error code conversion? */
+	errno = EINVAL;
+	return -1;
+}
+
+static int zstream_tls_flush(zstream stream)
+{
+	/* Believe or not, but mbedTLS doesn't buffer outbut data,
+	 * even for a single byte written it will create a TLS record
+	 * and send at once.
+	 */
+	return 0;
+}
+
+int zstream_tls_close(zstream stream)
+{
+	struct zstream_tls *self = (struct zstream_tls *)stream;
+	int tls_res = mbedtls_ssl_close_notify(&self->ssl);
+	/* We need to close underlying stream regardless of TLS close
+	 * notify status, or the stream will be leaked.
+	 */
+	int stream_res = zstream_close(self->sock);
+
+	if (tls_res == 0 && stream_res == 0) {
+		return 0;
+	}
+
+	if (stream_res == 0) {
+		/* TODO: Check which mbedTLS errors are possible and improve
+		 * mapping to POSIX codes.
+		 */
+		errno = EINVAL;
+	}
+
+	return -1;
+}
+
+static const struct zstream_api zstream_tls_api = {
+	.read = zstream_tls_read,
+	.write = zstream_tls_write,
+	.flush = zstream_tls_flush,
+	.close = zstream_tls_close,
+};
+
+static int zstream_mbedtls_ssl_send(void *ctx, const unsigned char *buf,
+			     size_t len)
+{
+	zstream sock = ctx;
+	ssize_t outlen = zstream_write(sock, buf, len);
+
+	if (outlen != -1) {
+		return outlen;
+	}
+
+	if (errno == EAGAIN) {
+		return MBEDTLS_ERR_SSL_WANT_WRITE;
+	}
+
+	/* Generic error */
+	return MBEDTLS_ERR_NET_SEND_FAILED;
+}
+
+static int zstream_mbedtls_ssl_recv(void *ctx, unsigned char *buf,
+				    size_t len)
+{
+	zstream sock = ctx;
+	ssize_t outlen = zstream_read(sock, buf, len);
+
+	if (outlen != -1) {
+		return outlen;
+	}
+
+	if (errno == EAGAIN) {
+		return MBEDTLS_ERR_SSL_WANT_READ;
+	}
+
+	/* Generic error */
+	return MBEDTLS_ERR_NET_RECV_FAILED;
+}
+
+// TODO: FIXME!
+static int null_entropy_func(void *data, unsigned char *output, size_t len) {
+	(void)data;
+	(void)output;
+	(void)len;
+	// enjoy random bytes
```

Comment:
Yeah, definitely needs to be fixed :-). Without entropy, the resulting connection can be trivially decoded.

Reply:
> Yeah, definitely needs to be fixed :-). Without entropy, the resulting connection can be trivially decoded.

Sure, fixed now. On Zephyr and if entropy subsystem is enabled, it is used. If entropy subsystem is not enabled, or when running on non-Zephyr (i.e. on a generic POSIX system), a big fat warning is printed that the communication is not secure.